---
Author: Rachel Payne 
title: "Assignment 5 - DESeq2"
output: html_document
---


For more detailed instructions, please read the assignment page in the course textbook. 


Use this block to source your `main.R`.

```{r setup, message=FALSE, warning=FALSE}
source('main.R')
```

## 1. Reading and subsetting the data from verse_counts.tsv and sample_metadata.csv

To begin, we will subset the counts matrix to only contain a subset of samples
found in the original data. We have provided you with the *full* counts matrix
of samples and a metadata file containing relevant experimental variables for
each sample. These files are named 'verse_counts.tsv' and 'sample_metadata.csv',
respectively. Define a function that reads from these two files to construct a
counts matrix and matching sample dataframe containing only samples belonging to
the vP0 and vAd timepoints. This should correspond to a total of 4 samples
(vP0_1, vP0_2, vAd_1, 'vAd_2). Store this counts matrix and the sample dataframe
in a SummarizedExperiments object. Ensure that the name of your counts matrix
stored as an assay in the SummarizedExperiment object is `counts`.

```{r, warning=FALSE, message=FALSE, echo=FALSE}
#Read in verse_counts.tsv and sample_metadata.csv and subset to the listed samples
#Store these dataframes in a SummarizedExperiment object and display the summary here
sum_exp <- make_se('verse_counts.tsv', 'sample_metadata.csv', c('vP0', 'vAd'))
sum_exp
```

Display the colData from your SummarizedExperiment object here:

```{r, echo=FALSE}
#Display the sample data from your SE object here to make sure the order is right 
colData(sum_exp)
```

## 2. [Running DESeq2](http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#differential-expression-analysis)

DESeq2 has a built-in function that directly uses a SummarizedExperiment object.
Define a function that takes your SummarizedExperiment and a design formula and
runs DESeq2 with standard parameters. In our case, we wish to determine what
genes are differentially expressed between our samples based on timepoint and
our model will simply be `~ timepoint` and we will use `vP0` as our reference
level. 

Return a *list* containing both the results from DESeq2 as a dataframe as well
as the DESeqDataSet object updated with various statistics in the course of the
analysis (referred to as the dds in the vignette)

```{r, run deseq2, echo=FALSE}
#Run your DESeq2 function here which returns a list of the dds and the results
#Assign the results and the dds from the returned list to their own variables 
#for ease of use
#Return both the results from DESeq2 as well as the DESeqDataSet object
res <- return_deseq_res(sum_exp, ~ timepoint)
deseq2_result <- res$res
dds <- res$dds

# deseq2_result


```
## 3. Annotating results to construct a labeled volcano plot

Using the results generated by DESeq2, convert it to a `tibble` and add one
additional column that denotes whether a gene is either: 1. Upregulated and
significant at a given padj threshold, 2.Downregulated and significant at a
given padj threshold, or 3. Not significant at a given padj threshold. __Name
this column `volc_plot_status`. Make sure the values for these labels are UP, DOWN,
and NS, respectively. __ Have your function take the DESeq2 results as well as a
user-defined threshold for padj and set it at .10.

```{r, label results tibble, echo=FALSE}
#Label the results with the column as specified above
#Display the summary of the tibble
labeled_results <- label_res(deseq2_result, .10)
labeled_results 
```

### Captions for figures

```{r code block for descriptive captions}

caption_4 <- 'Figure 1. Histogram of the raw p-values from all the genes discovered, verifying that the p-values computed display observations with a peak of values close to 0 as well as a roughly uniform distribution as the values approach 1.'
caption_5 <- 'Figure 2. Histogram of the log2FoldChanges for the differentially expressed genes with a p-adjusted value threshold < .10. Visualize the distribution of the up and down regulated genes produced from the experiment.'
caption_6 <- 'Figure 3. Scatter plot for normalized counts of DE genes. Intended to be used a way to make sure the analysis of DE genes is working as intended'
caption_7 <- 'Figure 4. Volcano plot of DE gene results. Statistically significant genes with low p-values will appear higher on the y-axis of the plot and they will be separated by the magnitude and directionality of their change.'
caption_9 <- 'Descriptive caption for part 9' 

```

## 4. Diagnostic plot of the raw p-values for all genes

Using the same results tibble from the previous function, plot a histogram of
the raw p-values for all of the genes discovered in the experiment. Avoid using
the base ggplot2 theme and make sure to label any appropriate axes and
components of the plot.

```{r, raw pval histogram, echo=FALSE, warning=FALSE, fig.cap = caption_4}
#Histogram of raw p-values for all discovered genes
raw_pval_hist <- plot_pvals(labeled_results)
raw_pval_hist

```

## 5. Plotting the LogFoldChanges for differentially expressed genes

Using the results tibble, make a histogram of the Log2FoldChange values for all
significant differentially expressed genes at a user-defined padj threshold and
set this threshold to .10. Avoid using the base ggplot2 theme and make sure to
label any appropriate axes and components of the plot.

```{r, log2fc DE genes histogram, echo=FALSE, fig.cap = caption_5}
#Histogram of log2FC values for significant DE genes with padj (FDR) < .10
logfc_plot <- plot_log2fc(labeled_results, 0.10)
logfc_plot
```

## 6. Plotting the normalized counts of differentially expressed genes

Extract out the DESeq2 normalized counts (by size factors) for the top ten
differentially expressed genes as ranked by ascending padj. Make a scatter / 
jitter plot of the counts and label each by their samplenames. 

```{r, plot of normalized counts, echo=FALSE, fig.cap = caption_6}
#Make a scatter/jitter plot of the norm counts for the top ten DE genes ranked by padj
plot_norm_counts <- scatter_norm_counts(labeled_results, dds, 10)
plot_norm_counts 



```

## 7. Volcano Plot to visualize differential expression results

Using the status column you generated in part 3, make a volcano plot of the
DESeq2 results and color the points appropriately. 

```{r, volcano plot, echo=FALSE, warning=FALSE, fig.cap = caption_7}
#Volcano Plot (scatter plot with log2FC on x-axis and -log10(padj) on y-axis)
#Color points (genes) by status defined in part 3
volcano_plot <- plot_volcano(labeled_results)
volcano_plot 

```

## 8. Running fgsea 

Perform a GSEA using a ranked list log2FoldChange values for all genes
discovered in the DESeq2 results against the C2 Canonical Pathways gene set
collection. Remember to convert the gene identifiers in our results to the
appropriate matching format found in the C2 gene sets. fgsea expects a named
vector of gene level statistics and the gene sets in the form of a named list.
As recommended, please set the minSize and maxSize parameters to 15 and 500,
respectively. Otherwise, run fgsea using default parameters and the `fgsea()`
function.

```{r, run fgsea using log2fc, warning=FALSE, message=FALSE, echo=FALSE}
#Run fgsea using a ranked list of descending log2FC against the C2 canonical
#pathways gene set
#Set minsize to 15 and maxsize to 500, leave the other parameters as defaults
# fgsea_results <- run_gsea(labeled_results, 'c2.cp.v7.5.1.symbols.gmt', 15, 500)
# fgsea_results

```

```{r, fgsea results, echo=FALSE}
#Display the results of the fgsea in a tibble sorted by padj (ascending)


## 9. Plotting the top ten positive NES and top ten negative NES pathways

#Plot the top ten pathways with both positive and negative NES (20 total)
#Color the pathways by the sign of their NES (positive or negative)


```

